In [111]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [112]:
df_bank = pd.read_csv('bank-additional-full-1.csv', sep=';')

# EDA
- Shape
- Info
- Describe for discriptive analysis (Helps in finding out the Outlier)
- Outlier Removal (Implicit null vlaue removal)
- Null analysis (If we remove the Outilers before then no need to run Null analysis)


Here is an explanation of the columns in the dataset:

**Client Data:**
*   `age`: Age of the client (numeric)
*   `job`: Type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
*   `marital`: Marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
*   `education`: Client's education level (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
*   `default`: Has credit in default? (categorical: 'no','yes','unknown')
*   `housing`: Has housing loan? (categorical: 'no','yes','unknown')
*   `loan`: Has personal loan? (categorical: 'no','yes','unknown')

**Last Contact Information:**
*   `contact`: Contact communication type (categorical: 'cellular','telephone')
*   `month`: Last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
*   `day_of_week`: Last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
*   `duration`: Last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

**Campaign Information:**
*   `campaign`: Number of contacts performed during this campaign and for this client (numeric, includes last contact)
*   `pdays`: Number of days that passed after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
*   `previous`: Number of contacts performed before this campaign and for this client (numeric)
*   `poutcome`: Outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')

**Social and Economic Context Attributes:**
*   `emp.var.rate`: Employment variation rate - quarterly indicator (numeric)
*   `cons.price.idx`: Consumer price index - monthly indicator (numeric)
*   `cons.conf.idx`: Consumer confidence index - monthly indicator (numeric)
*   `euribor3m`: Euribor 3 month rate - daily indicator (numeric)
*   `nr.employed`: Number of employees - quarterly indicator (numeric)

**Output Variable (Target):**
*   `y`: Has the client subscribed a term deposit? (binary: 'yes','no')

In [113]:
df_bank.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56.0,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57.0,services,married,high.school,unknown,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37.0,services,married,high.school,no,yes,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40.0,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56.0,services,married,high.school,no,no,yes,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [114]:
df_bank.shape

(41199, 21)

In [115]:
df_bank.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56.0,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57.0,services,married,high.school,unknown,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37.0,services,married,high.school,no,yes,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40.0,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56.0,services,married,high.school,no,no,yes,telephone,may,mon,...,1.0,999.0,0.0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [116]:
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41199 entries, 0 to 41198
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41195 non-null  float64
 1   job             41194 non-null  object 
 2   marital         41194 non-null  object 
 3   education       41194 non-null  object 
 4   default         41195 non-null  object 
 5   housing         41196 non-null  object 
 6   loan            41195 non-null  object 
 7   contact         41195 non-null  object 
 8   month           41196 non-null  object 
 9   day_of_week     41196 non-null  object 
 10  duration        41196 non-null  float64
 11  campaign        41196 non-null  float64
 12  pdays           41196 non-null  float64
 13  previous        41196 non-null  float64
 14  poutcome        41194 non-null  object 
 15  emp.var.rate    41196 non-null  float64
 16  cons.price.idx  41195 non-null  float64
 17  cons.conf.idx   41196 non-null 

In [117]:
df_bank.describe()

,age,duration,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed
count,41195.000000,41196.000000,41196.000000,41196.000000,41196.000000,41196.000000,41195.000000,41196.000000,41196.000000,41196.000000
mean,40.030319,258.281265,2.567676,962.482547,0.173124,0.081656,93.575867,-40.504600,3.620787,5166.996405
std,10.431818,259.254211,2.769751,186.893451,0.494987,1.570894,0.578999,4.629972,1.734655,72.300104
min,17.000000,0.000000,1.000000,0.000000,0.000000,-3.400000,92.201000,-50.800000,0.634000,4963.600000
25%,32.000000,102.000000,1.000000,999.000000,0.000000,-1.800000,93.075000,-42.700000,1.344000,5099.100000
50%,38.000000,180.000000,2.000000,999.000000,0.000000,1.100000,93.749000,-41.800000,4.857000,5191.000000
75%,47.000000,319.000000,3.000000,999.000000,0.000000,1.400000,93.994000,-36.400000,4.961000,5228.100000
max,98.000000,4918.000000,56.000000,999.000000,7.000000,1.400000,94.767000,-26.900000,5.045000,5228.100000


The descriptive analytics for each column provide insights into the data distribution:

*   **age**: The average age of clients is around 40, with a range from 17 to 98. The distribution is right-skewed **(Mean > Median)**, as seen by the median (38) being less than the mean. There are 4 null values.
*   **duration**: The average contact duration is about 258 seconds. This column has a large standard deviation (259.25), indicating high variability. The maximum duration is very high (4918 seconds), suggesting potential outliers. There are 3 null values.
*   **campaign**: The average number of contacts during this campaign is about 2.57, with a maximum of 56. This column is also right-skewed. The standard deviation is 2.77. There are 3 null values.
*   **pdays**: The average number of days since the last contact from a previous campaign is about 962. The high median and 75th percentile at 999 suggest that most clients were not contacted in previous campaigns. The standard deviation is 186.89. There are 3 null values.
*   **previous**: The average number of contacts before this campaign is about 0.17, with a maximum of 7. This indicates that most clients were not contacted in previous campaigns. The standard deviation is 0.49. There are 3 null values.
*   **emp.var.rate**: The average employment variation rate is close to zero. The range is from -3.4 to 1.4. The standard deviation is 1.57. There are 3 null values.
*   **cons.price.idx**: The consumer price index varies between 92.2 and 94.767, with an average around 93.58. The standard deviation is 0.58. There are 4 null values.
*   **cons.conf.idx**: The consumer confidence index ranges from -50.8 to -26.9, with an average around -40.5. The standard deviation is 4.63. There are 3 null values.
*   **euribor3m**: The Euribor 3-month rate has a wide range (0.634 to 5.045) and an average around 3.62. The standard deviation is 1.73. There are 3 null values.
*   **nr.employed**: The number of employees ranges from 4963.6 to 5228.1, with an average around 5167. The standard deviation is 72.30. There are 3 null values.
*   **job**: There are 5 null values.
*   **marital**: There are 5 null values.
*   **education**: There are 5 null values.
*   **default**: There are 4 null values.
*   **housing**: There are 3 null values.
*   **loan**: There are 4 null values.
*   **contact**: There are 4 null values.
*   **month**: There are 3 null values.
*   **day_of_week**: There are 3 null values.
*   **poutcome**: There are 5 null values.
*   **y**: There are 3 null values.

In [118]:
df_bank.isnull().sum()

,0
age,4
job,5
marital,5
education,5
default,4
housing,3
loan,4
contact,4
month,3
day_of_week,3


In [119]:
df_bank.describe(include="O")

,job,marital,education,default,housing,loan,contact,month,day_of_week,poutcome,y
count,41194,41194,41194,41195,41196,41195,41195,41196,41196,41194,41196
unique,12,4,8,3,3,3,2,10,5,3,2
top,admin.,married,university.degree,no,yes,no,cellular,may,thu,nonexistent,no
freq,10422,24934,12168,32595,21584,33957,26151,13769,8623,35563,36556


In [120]:
df_bank['y'].value_counts()

,count
y,
no,36556
yes,4640


# **Outlier Detection**

In [121]:
for col in df_bank.select_dtypes(include=np.number):
  Q1 = df_bank[col].quantile(0.25)
  Q3 = df_bank[col].quantile(0.75)
  IQR = Q3 - Q1
  lb = Q1 - 1.5 * IQR
  ub = Q3 + 1.5 * IQR
  df_bank = df_bank[(df_bank[col] >= lb) & (df_bank[col] <= ub)]

# **Target value imbalanced **

In [122]:
df_bank['y'].value_counts()

,count
y,
no,22709
yes,397


In [123]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import LabelEncoder

# Separate features (X) and target (y)
X = df_bank.drop('y', axis=1)
y = df_bank['y']

# Encode the target variable
le = LabelEncoder()
y_encoded = le.fit_transform(y)

# Select categorical and numerical columns
categorical_cols = X.select_dtypes(include='object').columns
numerical_cols = X.select_dtypes(include=np.number).columns

# Handle potential NaN values before one-hot encoding - fill with a placeholder or use handle_unknown='ignore'
# For simplicity, let's fill with 'missing' for object type columns
X[categorical_cols] = X[categorical_cols].fillna('missing')

# One-hot encode categorical features
X = pd.get_dummies(X, columns=categorical_cols, dummy_na=False)


# Apply SMOTE to the encoded data
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y_encoded)

# Display the value counts of the resampled target variable
print("Value counts after SMOTE:")
print(pd.Series(y_resampled).value_counts())

Value counts after SMOTE:
0    22709
1    22709
Name: count, dtype: int64


In [124]:
!pip install ydata

  Preparing metadata (setup.py) ... done
  Created wheel for ydata: filename=ydata-0.2-py3-none-any.whl size=3422 sha256=2cbe41875aebbda11918bef3f1a528ae350493410080d3cc71135d91b3671115
  Stored in directory: /root/.cache/pip/wheels/0a/e9/fb/3b84652c9f18b9038f8cdd6673f3cb22b4524b3754f1ec6b86
Successfully built ydata
